### PR TITLE
Round 3: Backup  VolumeSnapshots rather than larger DataMoverBackup

### PR DESCRIPTION
**Pull request by sseago**
This is still a bit rough, but I wanted to get something out to iterate on that reflects recent changes discussed.

### CODE REVIEW

File: docs/design/initial-design.md [hunk 1]
```diff
@@ -40,46 +40,33 @@ Here is a summary of the actions taken by the Velero CSI plugin on backup:
...
+### `DataMoverConfig`
+
+This could be implemented as a ConfigMap instead of a separate
+configuration CRD. In the absence of an explicit Backup resource, this
+will contain configuration necessary to tell the DataMover how to back
+up VolumeSnapshots as they are created. This will need to be either a
+well-known default name/namespace, or the Data Mover Controller will
+need an environment variable with a reference. In addition to
+configuring the plugin and storage location, this resource will also
+contain the selection specification for VolumeSnapshots to back up --
```
Comment (jwmatthews) [inline docs/design/initial-design.md:125]: possible typo?
'VolumeSnapshots to back up -- back up all snapshots'

File: docs/design/initial-design.md [hunk 2]
```diff
@@ -40,46 +40,33 @@ Here is a summary of the actions taken by the Velero CSI plugin on backup:
...
 with them.
 
-Here is an example of a `DataMoverBackup` resource.
+Open question: Do we need a separate StorageLocation CRD, or can that
+be merged into this Configuration resource?
 
-```
 apiVersion: oadp.openshift.io/v1alpha1
-kind: DataMoverBackup
+kind: DataMoverConfig
```
Comment (jwmatthews) [inline docs/design/initial-design.md:140]: The formatting for this block doesn't look right.
Looks like the markdown for code block was removed

File: docs/design/initial-design.md [hunk 3]
```diff
@@ -88,69 +75,55 @@ spec:
...
+
+We need a way to model a storage location for PVC storage. The
+simplest thing to do here would be just to reuse either Velero's
+`BackupStorageLocation` model, or CAM's `MigStorage` model. We could
+define our own with its own model, or we could even have a more
+flexible system where a `DataMoverStorageLocation` could optionally
+reference a Velero `BackupStorageLocation`, a CAM `MigStorage`, or a
+third data-mover-internal model. Regardless of the choice here, the
+metadata needs are approximately the same as the pre-existing Velero
+and CAM CRDs. In subsequent examples, `StorageLocation` is used as a
```
Comment (jwmatthews) [inline docs/design/initial-design.md:169]: Makes sense, main thought I have is we want to ensure DataMover is viable on it's own without Velero or CAM/MTC, the 'third data-mover-internal model' assume would suffice.

**Review by jwmatthews — CHANGES_REQUESTED** [review]
I struggled a bit to understand the high level approach, I think we should work to make this a little clearer for others.  

My suggestions are:
1. Add an explicit Goals section
1. Add an Approach summary up front, walk through high level Backup and Restore workflows
1. Add a picture to capture high level process for Backup, same for Restore
1. Add a section for 'Plugin Details' to describe the plugin requirements we envision.


Below are some other discussion points we should chat on:
* What do we want the plugin interface to be?  I am currently thinking a light weight interface that is essentially, Vendor provides an image and we will run it.  We will mount a PV(s) to it and pass in needed info, otherwise it's up to vendor to do what they like and use any technology they like.

* How do we go from a VolumeSnapshot is available to work happening?  Assume we will have a mechanism for configuring the controller to watch a certain type of VolumeSnapshot, perhaps through labels, and on matched VolumeSnapshots do work

* For Goals, I assume we want something like
   1. Copy data from a CSI snapshot to an external location via a plugin approach   
   1. Support both filesystem and block PV processing   
   1. Support working independent of Velero

* For Approach, I was thinking something like below considering 'Backup workflow'
  1. Create a configuration which instructs DataMover to watch VolumeSnapshots for a given criteria and where to copy the data to   
  1. For a matched VolumeSnapshot, create a new PV/PVC that is initiated from the referenced Snapshot
  1. Mount the PV on a Pod to handle the copying
      1. Assume the image the Pod runs is the 'backup plugin', perhaps a simple contract of mount the PV at a given location and mount the snapshot metadata at a specific file location in the container
      1. Communicate status/errors/progress via some method to be defined, maybe the container sends a curl command to update status of a specific CR for info


I'd like to get the approach/highlevel nailed down first, then we can flesh out the CRD/Controller details later.

File: docs/design/initial-design.md (hunk 3 above)
Comment (sseago) [reply docs/design/initial-design.md:169]: I was leaving this open-ended for the moment. What's not clear at this point is whether we want to require a data-mover-specific StorageLocation model (even if it ends up being a clone of either the velero or MTC storage model), or whether we create this clone storage model but *also* allow the user to specify storage via the velero/MTC models, or whether we may even want to refactor the MigStorage and DataMoverStorageLocation into a common model shared by both. From a development point of view a custom DataMoverStorageLocation that doesn't directly interact with velero or MTC models would be cleanest, but it puts more work on the user who already uses Velero and/or MTC to create separate DataMover resources.

But I think we're in complete agreement on the intent, that neither Velero nor MTC are required to use the DataMover.

File: docs/design/initial-design.md (hunk 1 above)
Comment (sseago) [reply docs/design/initial-design.md:125]: I'm not seeing a typo, but maybe the wording is confusing. Perhaps I should break it out into a bulleted list, but the point is there are at least 3 different ways to specify a collection of VolumeSnapshots:
1) all of them
2) all of them in a list of namespaces
3) all of them matching a LabelSelector

File: docs/design/initial-design.md (hunk 2 above)
Comment (sseago) [reply docs/design/initial-design.md:140]: ahh, yeah. not sure if it was removed or it was never there, but yeah. good catch.

**Comment by sseago** [comment]
@jwmatthews I've added those introductory sections. I think at this point, the major open questions that are referenced in the detailed design are also mentioned in the Goals, Approach, and Plugin Approach sections, so we can ignore the remainder until no major open questions remain in these sections.

**Review by jwmatthews — APPROVED** [review]

**Comment by jwmatthews** [comment]
LGTM to me.

I assume we'll iterate in future when we revisit, can answer some of the outstanding questions and tweak as we learn more.

**Pull request closed without merging**